chia/full_node/fee_tracker.py

```diff
@@ -62,12 +62,23 @@ def get_estimate_time_intervals() -> List[uint64]:
 
 
 def get_bucket_index(sorted_buckets: SortedDict, fee_rate: float) -> int:
+    if len(sorted_buckets) < 1:
+        raise RuntimeError("get_bucket_index: sorted_buckets is invalid")
+
     if fee_rate in sorted_buckets:
         bucket_index = sorted_buckets[fee_rate]
     else:
         # Choose the bucket to the left if we do not have exactly this fee rate
+        # Python's list.bisect_left returns the index to insert a new element into a sorted list
         bucket_index = sorted_buckets.bisect_left(fee_rate) - 1
```

Comment:
let's assume that the fee rate computed by a fee divided by cost is never exactly on the boundary of buckets (that seems like a fairly safe assumption, is it?). Any fee rate lower than `INITIAL_STEP` will return 0 here. i.e. the first bucket. Otherwise you get the bucket whose range you fall into. e.g. if the fee is between `INITIAL_STEP`and `INITIAL_STEP` + `STEP_SIZE`, you get bucket 1.

The edge case is if the fee rate is greater than the largest fee you can imagine (i.e. `INFINITE_FEE_RATE`), then you get back an index that doesn't exist, one past the end. (I believe that's what `bisect_left()` returns).

It seem to me the `-1` is suspect, and the check to see if `bucket_index` < 0 is suspect. The check to see if you're one past the end is unavoidable I think. This all assumes fee rates can't be negative of course, so maybe there should be an assert for that.

My assumption is that the fee "key" defines the upper bound of fees that belong in that bucket. But perhaps the intention is that the keys defined the lower bound, and that's why you subtract 1 from the bucket index. I would think a more intuitive way to do this would be to actually create the bottom bucket using the `MIN_FEE_RATE` constant (which is 0). I notice that this constant isn't used anywhere, suggesting that perhaps that was the intention.

Reply:
True, the `bucket_index < 0` check is only there because of the `-1`